common/include/pcl/pcl_macros.h

```diff
@@ -85,15 +85,16 @@
 // when compiling with some versions of nvcc so we
 // additionally check if nvcc is used before setting the
 // PCL_DEPRECATED_IMPL macro to [[deprecated]].
-#if defined(__has_cpp_attribute) && __has_cpp_attribute(deprecated) && !defined(__CUDACC__)
+// Furthermore MSVC 2017 has a bug which causes a C4067 warning
+// when using __has_cpp_attribute, so need first check for MSVC
+// before using __has_cpp_attribute to suppress this warning.
+// See: https://developercommunity.visualstudio.com/content/problem/327796/-has-cpp-attribute-emits-warning-is-wrong-highligh.html
+#if defined(_MSC_VER) && !defined(__CUDACC__)
```

Comment:
Since nvcc is the special case, we can reorder this to have some logic with nvcc and some other logic without nvcc. Like you mentioned, `[[deprecated]]` syntax should be the default for most compilers

Reply:
I have no idea if nvcc supports any command or not. So how do you ant it to have?

Reply:
First check if nvcc, if yes, fall back to `__attribute__` for GCC or Clang (and empty for MSVC) (original L91 and L96)
If not nvcc, then check for `[[deprecated]]` using `__has_cpp_attribute(deprecated)`. No check for `__has_cpp_attribute` required anymore. If no, complain about the compiler (L99). If yes, L93.

Reply:
This just make it worse:
```
#if defined __CUDACC__
  #if defined(__GNUC__) || defined(__clang__)
    #define PCL_DEPRECATED_IMPL(message) __attribute__((deprecated(message)))
  #else
    #define PCL_DEPRECATED_IMPL(message)
  #endif
#elif defined(_MSC_VER)
  #define PCL_DEPRECATED_IMPL(message) [[deprecated(message)]]
#elif defined(__has_cpp_attribute) && __has_cpp_attribute(deprecated)
  #define PCL_DEPRECATED_IMPL(message) [[deprecated(message)]]
#elif defined(__GNUC__) || defined(__clang__)
  #define PCL_DEPRECATED_IMPL(message) __attribute__((deprecated(message)))
#else
  #warning "You need to implement PCL_DEPRECATED_IMPL for this compiler"
  #define PCL_DEPRECATED_IMPL(message)
#endif
```

Reason: Due to the bug mentioned in my initial post we cannot check for MSVC and `__has_cpp_attribute(deprecated)` in the same line.

Reply:
Outside nvcc check, we don't need to check for `_MSC_VER` or `__GNUC__` or `__clang__`, since AFAIK, all other compilers are standard compliant

Reply:
```
#if defined(__CUDACC__)
  #if defined(__GNUC__) || defined(__clang__)
    #define PCL_DEPRECATED_IMPL(message) __attribute__((deprecated(message)))
  #else
    #define PCL_DEPRECATED_IMPL(message)
  #endif
#elif defined(_MSC_VER) || (__has_cpp_attribute(deprecated))
  #define PCL_DEPRECATED_IMPL(message) [[deprecated(message)]]
#else
  #warning "You need to implement PCL_DEPRECATED_IMPL for this compiler"
  #define PCL_DEPRECATED_IMPL(message)
#endif
```

Reply:
Wont work, as this does not fix the compiler bug. And btw: you can't even use `__has_cpp_attribute(deprecated)` without `defined(__has_cpp_attribute)` first, as this causes a `fatal error C1012: unmatched parenthesis: missing ')'` (check it on godbolt with MSVC 2017 and `/W3` as compiler parameter)

Reply:
[https://godbolt.org/z/a3SAae](Godbolt) with working warnings on nvcc 9.1 (linux), msvc 19.15, and ancient clang and gcc

The missing `__has_cpp_attribute` isn't an issue since we can are assuming that only MSVC and nvcc have issues with a standard macro. The error is coming because MSVC didn't implement a compliant pre-processor till 19.20